carapax-i18n/src/lib.rs

```diff
@@ -0,0 +1,243 @@
+//! An i18n handler for carapax
+
+#![warn(missing_docs)]
+
+use carapax::{core::types::Update, Context, HandlerFuture, HandlerResult, UpdateHandler};
+use gettext::Catalog;
+use std::sync::Arc;
+
+/// An i18n handler for carapax
+#[derive(Debug, Clone)]
+pub struct I18nHandler<R> {
+    resolver: R,
+    default_translator: Translator,
+    translators: Vec<Translator>,
+}
+
+impl<R> I18nHandler<R> {
+    /// Creates new I18nHandler
+    ///
+    /// Search translator in languages or use default one
+    pub fn new<I>(resolver: R, default_translator: Translator, translators: I) -> Self
+    where
+        I: IntoIterator<Item = Translator>,
+    {
+        Self {
+            resolver,
+            default_translator,
+            translators: translators.into_iter().collect(),
+        }
+    }
+}
+
+impl<R> UpdateHandler for I18nHandler<R>
+where
+    R: LocaleResolver,
+{
+    fn handle(&self, context: &mut Context, update: &Update) -> HandlerFuture {
+        let lang = self.resolver.resolve(update);
+        let translator = self
+            .translators
+            .iter()
+            .find(|translator| lang.as_ref().map(|lang| translator.locale() == lang).unwrap_or(false))
+            .cloned()
+            .unwrap_or_else(|| self.default_translator.clone());
+        context.set(translator);
+        HandlerResult::Continue.into()
+    }
+}
+
+/// An i18n translator uses `gettext` crate
+#[derive(Debug, Clone)]
+pub struct Translator {
+    locale: String,
+    catalog: Arc<Catalog>,
+}
+
+impl Translator {
+    /// Creates new translator
+    pub fn new<S: Into<String>>(locale: S, catalog: Catalog) -> Self {
+        Self {
+            locale: locale.into(),
+            catalog: Arc::new(catalog),
+        }
+    }
+
+    /// Translates string with given message ID
+    pub fn translate<M: Into<TranslationKey>>(&self, key: M) -> String {
+        let key = key.into();
+        let singular = &key.id;
+        match &key.kind {
+            TranslationKind::Singular => self.catalog.gettext(singular),
+            TranslationKind::Plural { id: plural, n } => self.catalog.ngettext(singular, plural, *n),
+        }
+        .to_string()
+    }
+
+    /// Returns locale
+    pub fn locale(&self) -> &str {
+        self.locale.as_str()
+    }
+}
+
+/// A locale resolver trait
+pub trait LocaleResolver {
+    /// Resolves locale and returns [Translator](Translator)
+    fn resolve(&self, update: &Update) -> Option<String>;
+}
+
+/// Resolves translator by user's language code
```

Comment:
Could you please elaborate a bit on why this empty struct is even needed? It's unclear from this short doc comment :(